aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -52,33 +66,44 @@ Tensor to(const Tensor& self, const TensorOptions& options, bool non_blocking, b
   if (dtype_opt) {
     specified_options = specified_options.dtype(dtype_opt.value());
   }
-  return to_impl(self, specified_options, non_blocking);
+  return to_impl(self, specified_options, non_blocking, optional_memory_format);
 }
 
-Tensor to(const Tensor& self, Device device, ScalarType dtype, bool non_blocking, bool copy) {
+Tensor to(const Tensor& self, Device device, ScalarType dtype, bool non_blocking, bool copy, c10::optional<c10::MemoryFormat> optional_memory_format) {
   device = ensure_has_index(device);
-  if (self.device() == device && self.dtype() == dtype && !copy) {
+  if (self.device() == device && self.dtype() == dtype && !copy &&
+      optional_memory_format.value_or(MemoryFormat::Preserve) ==
+          MemoryFormat::Preserve) {
     return self;
   }
-  return to_impl(self, self.options().device(device).dtype(dtype), non_blocking);
+  return to_impl(
+      self,
+      self.options().device(device).dtype(dtype),
+      non_blocking,
+      optional_memory_format);
 }
 
-Tensor to(const Tensor& self, ScalarType dtype, bool non_blocking, bool copy) {
-  if (self.dtype() == dtype && !copy) {
+Tensor to(const Tensor& self, ScalarType dtype, bool non_blocking, bool copy, c10::optional<c10::MemoryFormat> optional_memory_format) {
+  if (self.dtype() == dtype && !copy &&
+      optional_memory_format.value_or(MemoryFormat::Preserve) ==
+          MemoryFormat::Preserve) {
     return self;
   }
-  return to_impl(self, self.options().dtype(dtype), non_blocking);
+  return to_impl(
+      self, self.options().dtype(dtype), non_blocking, optional_memory_format);
 }
 
-Tensor to(const Tensor& self, const Tensor& other, bool non_blocking, bool copy) {
+Tensor to(const Tensor& self, const Tensor& other, bool non_blocking, bool copy, c10::optional<c10::MemoryFormat> optional_memory_format) {
   auto self_options = self.options();
   auto options = other.options();
   // Tensor.options() always have everything filled so we are happy and don't
   // even need to fill in device index.
-  if (self_options == options && !copy) {
+  if (self_options == options && !copy &&
+      optional_memory_format.value_or(MemoryFormat::Preserve) ==
+          MemoryFormat::Preserve) {
```

Comment:
Seeing that this logic is replicated many times, maybe it is worth pushing it into the `to_impl` implementation...